File: packages/server/src/services/pool-access.service.ts
import { PoolAccessInfo } from '@shared/types/poolAccessInfo';

const KYC_PATTERNS = [
  /\binstitutional\b/i,
  /\baccredited\b/i,
  /\bkyc\b/i,
  /\bwhitelist/i,
  /\bqualified\b/i,
  /\bpermissioned\b/i,
];

const LOCKUP_PATTERNS: Array<{ pattern: RegExp; description: (match: RegExpMatchArray) => string }> = [
  {
    pattern: /(\d+)\s*days?\s+unstaking/i,
    description: (m) => `${m[1]}-day unstaking period`,
  },
  {
    pattern: /unstaking\s+cooldown[:\s]+(\d+(?:\.\d+)?)\s*days?/i,
    description: (m) => `${m[1]}-day unstaking cooldown`,
  },
  {
    pattern: /(\d+)\s*d\s+(?:unlock|withdrawal\s+cycle)/i,
    description: (m) => `${m[1]}-day withdrawal cycle`,
  },
  {
    pattern: /(\d+)\s*days?\s+(?:lock|lockup|locked)/i,
    description: (m) => `${m[1]}-day lockup`,
  },
  {
    pattern: /(\d+)\s*(?:day|d)\s+cooldown/i,
    description: (m) => `${m[1]}-day cooldown`,
  },
  {
    pattern: /maturity\s+\d/i,
    description: () => 'Fixed-term (held to maturity)',
  },
];

/** Projects that are known DEX AMMs where a secondary-market swap exit is always possible. */
const SWAP_EXIT_PROJECTS = new Set([
  'uniswap-v2',
  'uniswap-v3',
  'uniswap-v4',
  'curve',
  'balancer',
  'balancer-v2',
  'balancer-v3',
  'sushiswap',
  'pancakeswap',
  'aerodrome',
  'velodrome',
  'camelot',
  'ramses',
  'thena',
  'trader-joe',
  'quickswap',
  'orca',
  'raydium',
  'pendle',
]);

/**
 * Projects that are known to require KYC / accreditation for entry.
 * This supplements `poolMeta` parsing for protocols that gate all pools.
 */
const KYC_ENTRY_PROJECTS = new Set(['maple', 'maple-v2', 'centrifuge', 'credix', 'goldfinch']);

function hasKycMeta(poolMeta: string | null): boolean {
  if (!poolMeta) {
    return false;
  }
  return KYC_PATTERNS.some((p) => p.test(poolMeta));
}

function detectLockup(poolMeta: string | null): string | null {
  if (!poolMeta) {
    return null;
  }
  for (const { pattern, description } of LOCKUP_PATTERNS) {
    const match = poolMeta.match(pattern);
    if (match) {
      return description(match);
    }
  }
  return null;
}

export function derivePoolAccessInfo(
  project: string,
  poolMeta: string | null,
  exposure: string | null,
): PoolAccessInfo {
  const kycFromMeta = hasKycMeta(poolMeta);
  const kycFromProject = KYC_ENTRY_PROJECTS.has(project.toLowerCase());
  const kycRequired = kycFromMeta || kycFromProject;

  const lockupDescription = detectLockup(poolMeta);
  const hasLockup = lockupDescription !== null;

  const canSwapToExit = SWAP_EXIT_PROJECTS.has(project.toLowerCase()) ? true : null;

  // A pool is illiquid when there is an explicit lockup period.
  // null means we have no data to determine liquidity.
  const isLiquid = hasLockup ? false : null;

  return {
    kycRequiredForEntry: kycRequired ? true : null,
    kycRequiredForExit: kycRequired ? true : null,
    canSwapToExit,
    isLiquid,
    lockupDescription,
  };
}

File: src/Credfeto.Defi.Server/Services/PoolAccessService.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Credfeto.Defi.Server.Models;

namespace Credfeto.Defi.Server.Services;

/// <summary>
///     Derives pool access and liquidity restriction information from pool metadata.
/// </summary>
internal static partial class PoolAccessService
{
    [GeneratedRegex(
        pattern: @"\binstitutional\b",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex InstitutionalPattern { get; }

    [GeneratedRegex(
        pattern: @"\baccredited\b",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex AccreditedPattern { get; }

    [GeneratedRegex(
        pattern: @"\bkyc\b",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex KycPattern { get; }

    [GeneratedRegex(
        pattern: @"\bwhitelist",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex WhitelistPattern { get; }

    [GeneratedRegex(
        pattern: @"\bqualified\b",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex QualifiedPattern { get; }

    [GeneratedRegex(
        pattern: @"\bpermissioned\b",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex PermissionedPattern { get; }

    [GeneratedRegex(
        pattern: @"(?<days>\d+)\s*days?\s+unstaking",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex LockupDaysUnstakingPattern { get; }

    [GeneratedRegex(
        pattern: @"unstaking\s+cooldown[:\s]+(?<days>\d+(?:\.\d+)?)\s*days?",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex LockupUnstakingCooldownPattern { get; }

    [GeneratedRegex(
        pattern: @"(?<days>\d+)\s*d\s+(?:unlock|withdrawal\s+cycle)",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex LockupWithdrawalCyclePattern { get; }

    [GeneratedRegex(
        pattern: @"(?<days>\d+)\s*days?\s+(?:lock|lockup|locked)",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex LockupDaysLockedPattern { get; }

    [GeneratedRegex(
        pattern: @"(?<days>\d+)\s*(?:day|d)\s+cooldown",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex LockupCooldownPattern { get; }

    [GeneratedRegex(
        pattern: @"maturity\s+\d",
        options: RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture,
        matchTimeoutMilliseconds: 500
    )]
    private static partial Regex MaturityPattern { get; }

    private static readonly HashSet<string> SwapExitProjects = new(StringComparer.OrdinalIgnoreCase)
    {
        "uniswap-v2",
        "uniswap-v3",
        "uniswap-v4",
        "curve",
        "balancer",
        "balancer-v2",
        "balancer-v3",
        "sushiswap",
        "pancakeswap",
        "aerodrome",
        "velodrome",
        "camelot",
        "ramses",
        "thena",
        "trader-joe",
        "quickswap",
        "orca",
        "raydium",
        "pendle",
    };

    private static readonly HashSet<string> KycEntryProjects = new(StringComparer.OrdinalIgnoreCase)
    {
        "maple",
        "maple-v2",
        "centrifuge",
        "credix",
        "goldfinch",
    };

    /// <summary>
    ///     Derives pool access info from the project identifier and pool metadata string.
    /// </summary>
    public static PoolAccessInfo DerivePoolAccessInfo(string project, string? poolMeta)
    {
        bool kycFromMeta = HasKycMeta(poolMeta);
        bool kycFromProject = KycEntryProjects.Contains(project);
        bool kycRequired = kycFromMeta || kycFromProject;

        string? lockupDescription = DetectLockup(poolMeta);
        bool hasLockup = lockupDescription is not null;

        bool? canSwapToExit = SwapExitProjects.Contains(project) ? true : null;
        bool? isLiquid = hasLockup ? false : null;

        return new PoolAccessInfo
        {
            KycRequiredForEntry = kycRequired ? true : null,
            KycRequiredForExit = kycRequired ? true : null,
            CanSwapToExit = canSwapToExit,
            IsLiquid = isLiquid,
            LockupDescription = lockupDescription,
        };
    }

    private static bool HasKycMeta(string? poolMeta)
    {
        return !string.IsNullOrEmpty(poolMeta)
            && (
                InstitutionalPattern.IsMatch(poolMeta)
                || AccreditedPattern.IsMatch(poolMeta)
                || KycPattern.IsMatch(poolMeta)
                || WhitelistPattern.IsMatch(poolMeta)
                || QualifiedPattern.IsMatch(poolMeta)
                || PermissionedPattern.IsMatch(poolMeta)
            );
    }

    private static string? DetectLockup(string? poolMeta)
    {
        if (string.IsNullOrEmpty(poolMeta))
        {
            return null;
        }

        Match m = LockupDaysUnstakingPattern.Match(poolMeta);

        if (m.Success)
        {
            return $"{m.Groups["days"].Value}-day unstaking period";
        }

        m = LockupUnstakingCooldownPattern.Match(poolMeta);

        if (m.Success)
        {
            return $"{m.Groups["days"].Value}-day unstaking cooldown";
        }

        m = LockupWithdrawalCyclePattern.Match(poolMeta);

        if (m.Success)
        {
            return $"{m.Groups["days"].Value}-day withdrawal cycle";
        }

        m = LockupDaysLockedPattern.Match(poolMeta);

        if (m.Success)
        {
            return $"{m.Groups["days"].Value}-day lockup";
        }

        m = LockupCooldownPattern.Match(poolMeta);

        if (m.Success)
        {
            return $"{m.Groups["days"].Value}-day cooldown";
        }

        return MaturityPattern.IsMatch(poolMeta) ? "Fixed-term (held to maturity)" : null;
    }
}
